### PR TITLE
Don't lint `irrefutable_let_patterns` on leading patterns if `else if` let-chains

### LINKED ISSUE #128661: `irrefutable_let_patterns` should not lint at the start of an `else if` let-chain
The following triggers `irrefutable_let_patterns`:

```rust
fn expensive() -> u32 {
    0
}

fn option(x: u32) -> Option<u32> {
    Some(x)
}

fn f(x: bool) {
    if x {
        // something
    } else if let foo = expensive()
        && let Some(bar) = option(foo)
    {
        // use `foo` and `bar`
    } else {
        // something
    }
}
```

Though `let foo = expensive()` is irrefutable, it would add an extra layer of nesting to move it into a regular `let` statement. e.g.

```rust
if x {
    // something
} else {
    let foo = expensive()
    if let Some(bar) = option(foo) {
        // use `foo` and `bar`
    } else {
        // something
    }
}
```

### PR DESCRIPTION
**Pull request by Jarcho**
fixes #128661

Is there any preference where the test goes? There looks to be several places it could fit.

### CODE REVIEW

**Comment by Nadrieril** [comment]
Could you add tests that reflect this change please?

**Comment by chenyukang** [comment]
I think `tests/ui/rfcs/rfc-2497-if-let-chains/` or `tests/ui/expr/if/` are both proper places for this test case?

**Comment by Jarcho** [comment]
Missed the `rfcs` folder.

File: tests/ui/rfcs/rfc-2497-if-let-chains/irrefutable-lets.rs [hunk 1]
```diff
@@ -75,4 +75,16 @@ fn main() {
         && let Range { start: local_start, end: _ } = first
         && let None = local_start {
     }
+
+    // No error. An extra nesting level would be required for the `else if`.
+    if opt == Some(None..None) {
+    } else if let x = opt.clone().map(|_| 1)
+        && x == Some(1)
+    {}
```
Comment (Nadrieril) [inline tests/ui/rfcs/rfc-2497-if-let-chains/irrefutable-lets.rs:83]: Could you add a test where the `else if` is split, e.g.
```rust
    if opt == Some(None..None) {
    } else {
        if let x = opt.clone().map(|_| 1)
            && x == Some(1)
        {}
    }
```

**Comment by Nadrieril** [comment]
Looks good to me! Thanks for taking the time to fix the little things :)

@bors r+

**Pull request merged**